scipy/sparse/csgraph/_tools.pyx

```diff
@@ -219,6 +219,61 @@ def csgraph_from_dense(graph,
                                                          infinity_null))
 
 
+def csgraph_from_adjacency_list(adjacency_list, *, weighted=True):
+    """Construct a CSR-format sparse graph from an adjacency list.
+
+    .. versionadded:: 1.8.0
+
+    Parameters
+    ----------
+    adjacency_list : dict
+        Input adjacency list. Could be either a dict of lists (for unweighted graph) or
+        a dict of dicts (for weighted graph). Dict length should be ``n_nodes``.
+    weighted : bool
+        If True (default), then construct a weighted graph.
+
+    Returns
+    -------
+    csgraph : csr_matrix
+        Compressed sparse representation of graph,
+
+    Examples
+    --------
+    >>> from scipy.sparse.csgraph import csgraph_from_adjacency_list
+
+    >>> graph = {
+    ...     0: {1: 1, 2: 2},
+    ...     1: {3: 1},
+    ...     2: {3: 3},
+    ...     3: {},
+    ... }
+
+    >>> csgraph_from_adjacency_list(graph)
+    <4x4 sparse matrix of type '<class 'numpy.int64'>'
+	    with 4 stored elements in Compressed Sparse Row format>
+
+    """
+    N = len(adjacency_list)
+
+    row = []
```

Comment:
Since `dict` limits us to pure-Python, chances are performance will never be amazing here, but a low-hanging fruit would be to target the `data`/`indices`/`indptr` initializer for `csr_matrix`: That one is more native and would be simpler to work with here: assuming all dictionaries are sorted by key, `data` is simply all your dict values stitched together, `indices` are the dict keys stitched together, and `indptr` is the cumulative sum of the lengths of the inner dictionaries.